components/camel-dynamic-router/src/main/java/org/apache/camel/component/dynamicrouter/DynamicRouterControlChannelProcessor.java

```diff
@@ -156,20 +157,22 @@ DynamicRouterControlMessage handleControlMessage(final Exchange exchange) {
     public boolean process(final Exchange exchange, final AsyncCallback callback) {
         LOG.debug("Received control channel message");
         DynamicRouterControlMessage controlMessage = handleControlMessage(exchange);
-        final DynamicRouterProcessor processor = Optional.ofNullable(component.getRoutingProcessor(controlMessage.getChannel()))
-                .orElseThrow(() -> new IllegalArgumentException(
-                        "Control channel message is invalid: wrong channel, or no processors present."));
-        switch (controlMessage.getMessageType()) {
-            case SUBSCRIBE:
-                processor.addFilter(controlMessage);
-                exchange.getIn().setBody(controlMessage.getId(), String.class);
-                break;
-            case UNSUBSCRIBE:
-                processor.removeFilter(controlMessage.getId());
-                break;
-            default:
-                // Cannot get here due to enum
-                break;
+        try (DynamicRouterMulticastProcessor processor
+                = Optional.ofNullable(component.getRoutingProcessor(controlMessage.getChannel()))
+                        .orElseThrow(() -> new IllegalArgumentException(
+                                "Control channel message is invalid: wrong channel, or no processors present."))) {
```

Comment:
This is wrong you should set exchange on exchange and not throw it. And then call callback done true and return true to exit

Reply:
Thanks.  I'll change it as requested.